### ISSUE TITLE
[Java API] Tensor.create() slow for large arrays

### ISSUE THREAD

**Issue by mdymczyk** [tensorflow/tensorflow#8244]
The current Java API's `Tensor.create(Object)` is really slow - for a batch of 128 images of size 224x224x3 it's taking around 1.5seconds. To put this into perspective `runner.run()` with that data and an InceptionV3 graph took below 1second so data prep is x1.5 of the runtime here (for a batch of 32 images it's around 0.35-0.45sec).

Is this working as intended? When running the Python code (using simple `sess.run(fetches, feed_dict=feed_dict)`) with which the graph meta file was generated (TF 1.0.1) and feeding a Python array I don't see such hiccups, the speed is the same as the Java `runner.run()`.

Might it be because of build flags used, maybe I'm missing some optimizations?

For now this small part is killing the whole performance, bringing it down from 130obs/sec (`runner.run()` time) to about ~45obs/sec (Tensor.create+run()).

A bit of a sidenote, the performance page states:

> This will result in poor performance.
> sess.run(train_step, feed_dict={x: batch_xs, y_: batch_ys})

But currently there's no other way to feed data from the Java API, right? A queue (able to read from a file and from memory, i.e. from a Java structure) would be amazing.

### Jar build command
```
export CC="/usr/bin/gcc"
export CXX="/usr/bin/g++"
export TF_NEED_CUDA=1
export GCC_HOST_COMPILER_PATH=$CC
export BUILDFLAGS="--config=cuda --copt=-m64 --linkopt=-m64 --copt=-march=native"

bazel build -c opt \
  //tensorflow/java:tensorflow \
  //tensorflow/java:libtensorflow_jni \
  $BUILDFLAGS --spawn_strategy=standalone --genrule_strategy=standalone
```

### Environment info
**OS:** Ubuntu 16.04
**GPU:** GPU TITAN X (Pascal) 12GB
**CPU:** Intel® Xeon® Processor E5-2630 v4 10core
**GPU Drivers:** 
NVidia CUDA Driver Version: 375.39
CUDNN 5.1.5
CUDA 8
**Tensorflow version:** JAR file built from current master (c25ecb53)

### Example

```java
public void test() {
  Random r = new Random();
  int imageSize = 224 * 224 * 3;
  int batch = 128;
  float[][] input = new float[batch][imageSize];
  for(int i = 0; i < batch; i++) {
    for(int j = 0; j < imageSize; j++) {
      input[i][j] = r.nextFloat();
    }
  }

  long start = System.nanoTime();
  Tensor.create(input);
  long end = System.nanoTime();
  // Around 1.5sec
  System.out.println("Took: " + (end - start));
}
```

**Comment by asimshankar** [comment]
Thanks for the detailed description and the sample code, it is very much appreciated!

The `create(Object)` method call involves use of reflection to determine the shape and copy things over one array at a time, so it is pretty slow, especially as you add dimensions. The [`create(shape, FloatBuffer)`](https://goo.gl/aGbcSB) method would be an order-of-magnitude faster. For example:

```java
public void test() {
    Random r = new Random();                                                                                                                 
    long[] shape = new long[] {batch, imageSize};                                                                                            
    FloatBuffer buf = FloatBuffer.allocate(imageSize * batch);                                                                               
    for (int i = 0; i < imageSize * batch; ++i) {                                                                                            
      buf.put(r.nextFloat());                                                                                                                
    }                                                                                                                                        
    buf.flip();                                                                                                                              
    long start = System.nanoTime();                                                                                                          
    Tensor.create(shape, buf);                                                                                                    
    long end = System.nanoTime();                                                                                                            
    System.out.println("Took: " + (end - start));                                                      
}
```

This is still slower than I'd want it to be, have to dig into that, but hopefully it is enough to satisfy your current needs (and session execution should be faster than Python).

Regarding your other question: Yes, using feeds is slower than getting input from queues. While I believe we do have the primitives to enable use of queues from any language, it is admittedly not too easy (as you have to figure out what is being done in python - e.g., start threads that run the enqueue op - and duplicate that). That is something we have to look into and contributions are welcome!

Do let me know if using the `FloatBuffer` suffices for now (and I can close this issue, while we look into general performance improvements for the Java API).

**Comment by mdymczyk** [comment]
@asimshankar this is really great stuff, the timings improved exactly by an order of magnitude as you said, thanks! A batch of 128 takes around 0.2s and the batch of 32 takes around 0.04s. Still I think a bit slower than what Python does but definitely I can work with that, thanks!

**Comment by AndreyFilippov** [comment]
@asimshankar, is there any way to reuse tensors of the same shape by just copying data without reallocating memory (maybe handling endiness outside)? Eventually we would like to feed directly from the non-tensorflow GPU memory (that seems recently to become possible - https://github.com/tensorflow/tensorflow/commit/a1d6179adb1ca6208281ed955860c319525edf75) but for now we would like to eliminate all unnecessary memory copying - the source float[] data is created by a multi-threaded application. Without it the inference takes less time than source data copying (even in Python).

**Comment by asimshankar** [comment]
At the moment, no.

Off the top of my head, one option would be to add an `alias(Buffer buffer, DataType dtype, long[] shape)` factory function to the `Tensor` class. If that works, a contribution will be welcome (@karllessard may be interested).

Note that there is a [private `Tensor.buffer()`](https://github.com/tensorflow/tensorflow/blob/f4de7ec889311c42b3af4d5f34f7d31f56f73177/tensorflow/java/src/main/java/org/tensorflow/Tensor.java#L572) method that provides a `ByteBuffer` view of the underlying memory. I'm not inclined to make it public as that can make it easy to run into memory safety concerns (e.g., the returned `ByteBuffer` object shouldn't outlive the `Tensor` object)

**Comment by AndreyFilippov** [comment]
Thanks for the suggestion. I'll try everything as is first, just keeping in mind the memory copying and your recommendation, and then try to implement it (or just try to jump over and go directly to feed from the GPU memory).
BTW - here is the project where we need fast feed (just converted the trained network to have stage2 fully convolutional) - https://blog.elphel.com/2018/09/neural-network-doubled-effective-baseline-of-the-stereo-camera/

**Comment by karllessard** [comment]
FYI, I’m experimenting on eager execution of Java operations right now and I’m already taking a look on how we can make a better usage of ˋTensor.buffer()ˋ for that purpose. @AndreyFilippov , I can keep you posted of my progress or please do so if you plan to implement something around it, thanks!

**Comment by theceday** [comment]
@asimshankar, @karllessard 
Could you please provide a basic example in which order we should put the data for FloatBuffer?

This one works, but causes wrong predictions.

`float[][][][] rawData = new float[BATCH_SIZE][IMAGE_WIDTH][IMAGE_HEIGHT][1];
..
..
return Tensor.create(rawData, Float.class);`

And there is a similar issue here, i guess:
https://github.com/tensorflow/tensorflow/issues/19359

So we need to use FloatBuffer directly I guess, official example is doing all stuff under the hood.
Could you clearify how do feed such multidimensional array?

I tried different possible version of putting data to FloatBuffer, code works, outputs becomes "completely" garbage then.

**Comment by Docjones** [comment]
We are having a similar problem: We create Tensors in java using `Tensor.<String>create(...);` We are feeding a `byte[][][]` containing padded Strings (see our [stackoverflow ](https://stackoverflow.com/questions/55296464/how-can-i-feed-a-sparse-placeholder-in-a-tensorflow-model-from-java)post regarding this).

We are feeding 4 Tensors with ~7500 Strings each, which quite some time to `create`. Has anyone a hint how we can improve the creation?

**Comment by karllessard** [comment]
@Docjones , did you tried something like this? Note that this will pad your `refNames` not with `\` but with zeros but I'm curious to see if that works.

Warning! I haven't ran the code before myself so it might require some adjustments...

```java
    final String TEST_NAME = "Max Mustermann";

    final String[] REF_NAMES = {
        "joseph haydn",
        "max mustermann",
        "erika musterfrau",
        "johann sebastian bach",
        "wolfgang amadeus mozart"
    };

    SavedModelBundle model = SavedModelBundle.load("/home/karl/dev/projects/tf/tests/dist_model", "serve");
    try (Graph g = new Graph()) {
      g.importGraphDef(model.metaGraphDef());
      
      ByteBuffer testInputData = ByteBuffer.wrap(TEST_NAME.getBytes("UTF-8"));
      
      byte[][] refNamesBytes = new byte[REF_NAMES.length][];
      int refNameBytesMaxLength = 0;

      // Convert the ref names into bytes and keep track of the longest sequence length
      for (int i = 0; i < REF_NAMES.length; ++i) {
        byte[] refNameBytes = REF_NAMES[i].getBytes("UTF-8");
        if (refNameBytesMaxLength < refNameBytes.length) {
          refNameBytesMaxLength = refNameBytes.length;
        }
        refNamesBytes[i] = refNameBytes;
      }

      // Serialize ref names to the buffer
      ByteBuffer refInputData = ByteBuffer.allocate(REF_NAMES.length * refNameBytesMaxLength);
      for (byte[] refNameBytes : refNamesBytes) {
        refInputData.put(refNameBytes);
        refInputData.position(refInputData.position() + (refNameBytesMaxLength - refNameBytes.length));
      }

      try (Tensor<String> testInput = Tensor.create(String.class, new long[0], testInputData);
          Tensor<String> refInput = Tensor.create(String.class, new long[] { REF_NAMES.length }, refInputData )) {
        List<Tensor<?>> outputs = model.session().runner()
            .feed("test_name", testInput)
            .feed("ref_names", refInput)
            .fetch("min_idx")
            .run();
        System.out.println("Nearest distance: " + outputs.get(0).longValue());
      }
```

**Comment by Docjones** [comment]
@karllessard - thanks for your help! We adapted your example to our (much bigger code) and unfortunately, we receive the following exception upon calling `model.session.runner.feed(...)`;

`Caused by: java.lang.IllegalArgumentException: Malformed TF_STRING tensor; too short to hold number of elements`

Most likely, the created ByteBuffer is different (not only in padding with `0` instead of `/`) from our self created byte[][][]. We can't see, what we have done wrong...

**Comment by karllessard** [comment]
Mmh, just an idea like this, maybe you should `ByteBuffer.rewind()` both buffers before feeding them to the session?

**Comment by rongsuema** [comment]
Hi all, any update on Docjones's problem? We are also working on this issue as we want to use the same idea, i.e. FloatBuffer API, to improve the performance of `Tensor<String>... Tensor.create(...)`. However, I just look into [another similar post] (https://github.com/tensorflow/tensorflow/issues/24331) , and find that @asimshankar suggest to use the API without buffer, which means we can't make that improvement in the same way in this case?

**Comment by karllessard** [comment]
I’ve dug a bit further and I saw that TF_String buffers are exceptional because they expect some extra bytes at the beginning of each elements, something called an “offset”.

I’ll continue to investigate when I’ll have a few minutes but it is something you might also want to look at if you’re curious to find before.

**Comment by karllessard** [comment]
Ok, yeah it looks like the way to encode a TF string in a buffer is way more complex that I was first thought :-/ Basically, it follows as structure like this (for String array `[A, B, C]`):

`[offset A][offset B][offset C][length A][data A][length B][data B][length C][data C]`

After a few tests, I think the following code should do the job, I'm curious that you give it a try in your projects and if 1) it works and 2) it is faster, I can push it as a utility in the repo.

```java
  private static ByteBuffer stringArrayToBuffer(String[] values) throws IOException {
    long offsets[] = new long[values.length];
    byte[][] data = new byte[values.length][];
    int dataSize = 0;

    // Convert strings to encoded bytes and calculate required data size, including a varint for each of them
    for (int i = 0; i < values.length; ++i) {
      byte[] byteValue = values[i].getBytes("UTF-8");
      data[i] = byteValue;
      int length = byteValue.length + varintLength(byteValue.length);
      dataSize += length;
      if (i < values.length - 1) {
        offsets[i + 1] = offsets[i] + length;
      }
    }

    // Important: buffer must follow native byte order
    ByteBuffer buffer = ByteBuffer.allocate(dataSize + (offsets.length * 8)).order(ByteOrder.nativeOrder());

    // First, write offsets to each elements in the buffer
    for (int i = 0; i < offsets.length; ++i) {
      buffer.putLong(offsets[i]);
    }
    
    // Second, write strings bytes, each preceded by its length encoded as a varint
    for (int i = 0; i < data.length; ++i) {
      encodeVarint(buffer, data[i].length);
      buffer.put(data[i]);
    }

    return (ByteBuffer)buffer.rewind();
  }
  
  private static void encodeVarint(ByteBuffer buffer, int value) {
      int v = value;
      while (v >= 0x80) {
        buffer.put((byte)((v & 0x7F) | 0x80));
        v >>= 7;
      }
      buffer.put((byte)v);
  }

  private static int varintLength(int length) {
    int len = 1;
    while (length >= 0x80) {
      length >>= 7;
      ++len;
    }
    return len;
  }
```

**Comment by rongsuema** [comment]
Thanks a lot @karllessard :D ! I will have a try and get back to you once we have a result.

**Comment by Docjones** [comment]
Extraordinary, @karllessard. We will try that out - give us a bit. We are just launching a new service at the beginning next week and need to shift the focus a bit.

**Comment by dirkzhang** [comment]
@karllessard instead of using String[], I use org.tensorflow.example, how can I use Tensor<String> inputBatch = Tensor.create(String.class, shap, ByteBuffer) to improve response ? 
I just use org.tensorflow.example as one string, like that [offset A][length A][data A], it give me the same error java.lang.IllegalArgumentException: Malformed TF_STRING tensor; too short to hold number of elements

**Comment by karllessard** [comment]
@dirkzhang , not sure I understand correctly, what/where is `org.tensorflow.example` exactly?

Now the `String[]` here is not the data for a rank-1 String tensor but a flat view of all strings of your tensor, no matter its shape, like we do with other datatypes (Int, Floats...). The values are fed in the buffer starting iterating from dim N-1 towards dim 0.

For example, to represent a 3D matrix of strings like this one:

{ { {"000", "001", "002" }, { "010", "011", "012" } }, { { "100", "101", "102" }, { "110", "111", "112" } } }

the string array passed in parameter would be

{ "000", "001", "002", "010", "011", "012", "100", "101", "102", "110", "111", "112" }

and the shape [ 2, 2, 3 ]

Does this answer your question?

**Comment by rongsuema** [comment]
Hi @karllessard , me and @dirkzhang are working on this stuff, and sorry to make that confusion. 
The org.tensorflow.example.Example refers to the tf.Example which fundamentally a {"feature key": tf.train.Feature} mapping.

Actually, what we are doing here is that we use the tf.Example (i.e. org.tensorflow.example.Example in Java API) as the input of the prediction inference. It is kind of a Map instead of a List, and may look like this if we decode the tf.Example message:
`features {
  feature {
    key: "feature0"
    value {
      int64_list {
        value: 0
        value: 1
        value: 2
      }
    }
  }
  feature {
    key: "feature1"
    value {
      int64_list {
        value: 4
        value: 5
      }
    }
  }
  feature {
    key: "feature2"
    value {
      bytes_list {
        value: "goat"
      }
    }
  }
  feature {
    key: "feature3"
    value {
      float_list {
        value: 0.332442
        value: 0.1123344
      }
    }
  }
}` 

For now, we use `Tensors.create(new byte[][]{example.toByteArray()})` to create the Tensor<String> and obviously we can't simply use the length of the byteArray for the new 'create()' API. Also, representing example as only one string is not working as that @dirkzhang just did. 

So if I understand your point correctly, we need to figure out how many strings are in this example object. Is that right? Do you have any idea about this?

**Comment by karllessard** [comment]
Oh, the proto! Sorry for the confusion as well,

But before going any further with this approach, did you tried to add a [ParseExample](https://static.javadoc.io/org.tensorflow/libtensorflow/1.13.1/org/tensorflow/op/core/ParseExample.html) node to your graph instead (or the equivalent in Python)? I think that would be a proper way to extract features from an example protobuf. unless you need to do some preprocessing.

**Comment by rongsuema** [comment]
You are right, we need to do some preprocessing before we build those features and then add them to the Example object.

**Comment by karllessard** [comment]
Ok, so you want to create a tensor which is a batch of `Example` protobuf serialized as strings?

The code above should work, where each element of the `String[]` is a serialized example. The error you receive make me think that there might be a mismatch with the length of this array and the shape you are providing. I guess it should be either `new long[] { batchSize }` or `new long[] { batchSize, 1 }`, depending on your model, where `batchSize` is equal to `String[].length`.

For example:
`Tensor.create(String.class, new long[] { 1 }, stringArrayToBuffer(new String[] { example.toByteString() }))` 

Is it what you tried? I guess we can continue debugging using the developers mailing list](developers@tensorflow.org), I don't know if you are subscribed to it.

**Comment by rongsuema** [comment]
> so you want to create a tensor which is a batch of Example protobuf serialized as strings?
Exactly. And we might do something wrong... well, I will have a quick try. 

For the developers mailing list, will my email automatically  be assigned to you?

**Comment by karllessard** [comment]
> For the developers mailing list, will my email automatically be assigned to you?

I'll reply to you but your message will gain more visibility so we might get some input/help from other devs as well.

**Comment by rongsuema** [comment]
emm, I just sent an email to developers@tensorflow.org and got an automatic response from tensorflow.org admins:

> We're writing to let you know that the group you tried to contact (developers) may not exist, or you may not have permission to post messages to the group. A few more details on why you weren't able to post

please let me know if you can find that email.

**Comment by karllessard** [comment]
ok, so please write me directly to my personal email, we might lack the extra-visibility but it will be more interactive: karl at kubx dot ca

**Comment by karllessard** [comment]
For those who were following the thread, I dropped a demo (draft) of how an example.proto could be serialized in a `ByteBuffer` and parsed in a graph written in Java [here](https://github.com/karllessard/models/tree/master/samples/languages/java/mnist/src/main/java/org/tensorflow/model/sample/buffer)

**Comment by maziyarpanahi** [comment]
Hi @karllessard, the link you shared that has the example of ByteBuffer is a broken link now. Could it be possible to re-share that? Many thanks.

**Comment by karllessard** [comment]
@maziyarpanahi : sure, it's now here: https://github.com/karllessard/tensorflow-models/tree/master/samples/languages/java/suite/src/main/java/org/tensorflow/model/sample/buffer

**Comment by karllessard** [comment]
By the way, all this is currently being addressed in the [new official Java repository for TensorFlow](https://github.com/tensorflow/java), there is a important project called "Tensor NIO" that I'm working on which will allow users to directly access tensor memory from Java and to read/write its data in an N-dimensional space. That will include what is found in the `Buffers` class of the previous example.

If anyone is interested to know more about it, just let me know.
